pkg/controller/jobs/raycluster/raycluster_controller.go

```diff
@@ -0,0 +1,198 @@
+	// len = workerGroups + head
+	podSets := make([]kueue.PodSet, len(j.Spec.WorkerGroupSpecs)+1)
+
+	// head
+	podSets[0] = kueue.PodSet{
+		Name:     headGroupPodSetName,
+		Template: *j.Spec.HeadGroupSpec.Template.DeepCopy(),
+		Count:    1,
+	}
+
+	// workers
+	for index := range j.Spec.WorkerGroupSpecs {
+		wgs := &j.Spec.WorkerGroupSpecs[index]
+		replicas := int32(1)
+		if wgs.Replicas != nil {
+			replicas = *wgs.Replicas
+		}
+		podSets[index+1] = kueue.PodSet{
+			Name:     strings.ToLower(wgs.GroupName),
+			Template: *wgs.Template.DeepCopy(),
+			Count:    replicas,
+		}
+	}
+	return podSets
+}
+
+func (j *RayCluster) RunWithPodSetsInfo(podSetsInfo []podset.PodSetInfo) error {
+	expectedLen := len(j.Spec.WorkerGroupSpecs) + 1
+	if len(podSetsInfo) != expectedLen {
+		return podset.BadPodSetsInfoLenError(expectedLen, len(podSetsInfo))
+	}
+
+	j.Spec.Suspend = ptr.To(false)
+
+	// head
+	headPod := &j.Spec.HeadGroupSpec.Template
+	info := podSetsInfo[0]
+	if err := podset.Merge(&headPod.ObjectMeta, &headPod.Spec, info); err != nil {
+		return err
+	}
+
+	// workers
+	for index := range j.Spec.WorkerGroupSpecs {
+		workerPod := &j.Spec.WorkerGroupSpecs[index].Template
+
+		info := podSetsInfo[index+1]
+		if err := podset.Merge(&workerPod.ObjectMeta, &workerPod.Spec, info); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (j *RayCluster) RestorePodSetsInfo(podSetsInfo []podset.PodSetInfo) bool {
+	if len(podSetsInfo) != len(j.Spec.WorkerGroupSpecs)+1 {
+		return false
+	}
+
+	changed := false
+	// head
+	headPod := &j.Spec.HeadGroupSpec.Template
+	changed = podset.RestorePodSpec(&headPod.ObjectMeta, &headPod.Spec, podSetsInfo[0]) || changed
+
+	// workers
+	for index := range j.Spec.WorkerGroupSpecs {
+		workerPod := &j.Spec.WorkerGroupSpecs[index].Template
+		info := podSetsInfo[index+1]
+		changed = podset.RestorePodSpec(&workerPod.ObjectMeta, &workerPod.Spec, info) || changed
+	}
+	return changed
+}
+
+func (j *RayCluster) Finished() (metav1.Condition, bool) {
+	condition := metav1.Condition{
+		Type:    kueue.WorkloadFinished,
+		Status:  metav1.ConditionFalse,
+		Reason:  string(j.Status.State),
+		Message: string(j.Status.Reason),
+	}
+	// Technically a RayCluster is never "finished"
+	return condition, false
+}
+
+func (j *RayCluster) PodsReady() bool {
+	return j.Status.State == rayv1.Ready
+}
+
+func SetupIndexes(ctx context.Context, indexer client.FieldIndexer) error {
+	return jobframework.SetupWorkloadOwnerIndex(ctx, indexer, gvk)
+}
+
+func GetWorkloadNameForRayCluster(jobName string) string {
+	return jobframework.GetWorkloadNameForOwnerWithGVK(jobName, gvk)
+}
+
+func isRayCluster(owner *metav1.OwnerReference) bool {
+	return owner.Kind == "RayCluster" && strings.HasPrefix(owner.APIVersion, "ray.io/v1")
+}
+
+func IsOwnerRayJob(r *RayCluster) bool {
+	if owner := metav1.GetControllerOf(r.Object()); owner != nil {
+		return owner.Kind == "RayJob" || jobframework.IsOwnerManagedByKueue(owner)
```

Comment:
Is `|| jobframework.IsOwnerManagedByKueue(owner)` actually needed here?

Reply:
Actually, it could be the other way, we just need to check `jobframework.IsOwnerManagedByKueue(owner)`.
Then in the future we don't need to change this line if there is another CRD managed by Kueue that can create RayClusters.

Reply:
Right. Could that logic be generalised and applicable to any integration then?

Reply:
It could. We can make it part it the job framework webhook helpers. But I'm fine leaving such thing for a follow up. Just matching the implementation for now should be enough.